neural_compressor/adaptor/torch_utils/waq/auto_alpha.py

```diff
@@ -0,0 +1,659 @@
+        self.op_types = op_types
+        self.absorb_to_layer = absorb_to_layer
+        self.weight_scale_dict = {}
+        self.q_func = q_func
+        self.example_inputs = example_inputs
+        self.max_value_info = {}  # to record max values for alpha tune
+        self.record_max_info = record_max_info[0] if isinstance(record_max_info, tuple) else record_max_info
+        self.weight_clip = weight_clip[0] if isinstance(weight_clip, tuple) else weight_clip
+        self.input_maxes = {}
+        self.input_mins = {}
+        self.input_maxes_abs = {}
+        self.device = device
+
+    def tune(self):
+        """The main entry of auto_alpha
+        :return: Optimal alpha values and scales based on user-defined recipes."""
+        calib = Calibration(self.model, self.dataloader, self.q_func, self.device)
+        calib_iter = 100
+        self.input_mins, self.input_maxes = calib.calibrate(calib_iter, self.op_types)
+        for key in self.input_mins.keys():
+            self.input_maxes_abs[key] = torch.max(torch.abs(self.input_mins[key]), torch.abs(self.input_maxes[key]))
+
+        scale_memo_use = 0
+        for key in self.absorb_to_layer:
+            layer_name = self.absorb_to_layer[key][0]
+            input_max = self.input_maxes_abs[layer_name]
+            scale_memo_use += 4 * input_max.shape[0] * len(self.absorb_to_layer[key])
+        alpha_space_len = (self.alpha_max - self.alpha_min) / self.alpha_step + 1
+        scale_memo_use *= alpha_space_len
+        self._save_scale = enough_memo_store_scale(self.device, scale_memo_use)
+
+        if self.loss_type == "blockwise":
+            self.block_names = self.get_blocks()
+            logger.info("Blockwise auto-tuning will be performed")
+            module_names = self._get_sq_layer_names()
+            block_names, self.block_to_module = self.block_names, {}
+            for block in block_names:
+                self.block_to_module[block] = []
+            for module in module_names:
+                checked = False
+                for block in block_names:
+                    if block + "." in module:
+                        self.block_to_module[block].append(module)
+                        checked = True
+                if not checked:
+                    self.block_to_module[module] = [module]
+            self.block_names = list(self.block_to_module.keys())
+            logger.info(f"Blockwise auto-tuning: {len(self.block_names)} blocks found")
+            logger.debug(f"Blockwise auto-tuning blocks info: {self.block_to_module}")
+            return self._auto_tune_alpha_blockwise()
+        else:
+            return self._auto_tune_alpha()
+
+    def get_blocks(self):
+        """Obtain a list of blocks in block-wise tuning mode."""
+        block_names = []
+        for n, m in self.model.named_modules():
+            if hasattr(type(m), "__name__") and "ModuleList" in type(m).__name__:
+                for nn, mm in m.named_children():
```

Comment:
this has bug for mixstral moe model, please refer to autoround. Besides, as sq needs to support mutlimodal models, please test a multimodal later.